### PR TITLE
fix(): error context for git_fetch refspec not found

### PR DESCRIPTION
**Pull request by dacianpascu06**
### What does this PR try to resolve?

Fixes Issue : #14621 

Adds more error context for fetching a commit that doesn't exists.

### How should we test and review this PR?

I've created two tests, one for fast_path and one for libgit2.

r? @weihanglo

### CODE REVIEW

File: tests/testsuite/git.rs [hunk 1]
```diff
@@ -4052,6 +4052,90 @@ src/lib.rs
         .run();
 }
 
+#[cargo_test(public_network_test)]
```
Comment (weihanglo) [inline tests/testsuite/git.rs:4055]: You'll need `requires_git` as well

```suggestion
#[cargo_test(public_network_test, requires_git)]
```

File: src/cargo/sources/git/utils.rs [hunk 2]
```diff
@@ -1152,6 +1152,13 @@ pub fn fetch(
                         continue;
                     }
                 }
+                if matches!(err.class(), ErrorClass::Net) && err.code() == git2::ErrorCode::Eof {
+                    let revspec_id: String = refspecs
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1156]: I would avoid doing this re-parsing, especially when we need to do that for every kind of fetch approaches (like, we've covered git CLI and `libgit2` here but missed gitoxide).

What do you think of doing it one level up as [mentioned here](https://github.com/rust-lang/cargo/issues/14621#issuecomment-2397279846), so we don't need to repeat ourselves for git/libgit2/gitoxide?

File: tests/testsuite/git.rs
Comment (weihanglo) [inline tests/testsuite/git.rs:1]: If possible, please follow [this guide](https://doc.crates.io/contrib/process/working-on-cargo.html#submitting-a-pull-request), which one commit contains a test showing the current problematic behavior with assertions, followed by the other commits fixing the behavior and test. Each commit is expected to pass CI.

File: src/cargo/sources/git/utils.rs [hunk 3]
```diff
@@ -1221,9 +1228,21 @@ fn fetch_with_cli(
...
-    cmd.exec()?;
+
+    let revspec_id: String = refspecs
+        .iter()
+        .map(|s| s.split(':').next().unwrap_or("").to_string())
+        .collect();
+
+    cmd.exec().with_context(|| {
+        format!(
+            "Failed to fetch via Github fast path \nrevspec {:?} not found",
```
Comment (ehuss) [inline src/cargo/sources/git/utils.rs:1242]: BTW, this does not look related to the GitHub fast path. I'm not sure I understand why it is mentioning it here.

File: src/cargo/sources/git/utils.rs [hunk 4]
```diff
@@ -944,9 +944,15 @@ pub fn fetch(
...
     let shallow = remote_kind.to_shallow_setting(repo.is_shallow(), gctx);
 
+    // Flag to keep track if the rev is a full commit hash
+    let mut fast_path_rev: bool = false;
+
     let oid_to_fetch = match github_fast_path(repo, remote_url, reference, gctx) {
         Ok(FastPathRev::UpToDate) => return Ok(()),
-        Ok(FastPathRev::NeedsFetch(rev)) => Some(rev),
+        Ok(FastPathRev::NeedsFetch(rev)) => {
+            fast_path_rev = true;
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:953]: Instead of adding a flag here, we should probably add one in `} else if let Some(oid_to_fetch) = oid_to_fetch {` after line 995, which the `ref/commit/{0}` really happened and caused the problem

File: tests/testsuite/git.rs [hunk 5]
```diff
@@ -4052,6 +4052,94 @@ src/lib.rs
...
+  failed to clone into: [ROOT]/home/.cargo/git/db/bitflags-[HASH]
+
+Caused by:
+  process didn't exit successfully: `git fetch --no-tags --force --update-head-ok 'https://github.com/rust-lang/bitflags.git' '+11111b376b93484341c68fbca3ca110ae5cd2790:refs/commit/11111b376b93484341c68fbca3ca110ae5cd2790'` ([EXIT_STATUS]: 128)
+
+"#]])
+        .run();
+}
+
+#[cargo_test(public_network_test, requires_git)]
```
Comment (weihanglo) [inline tests/testsuite/git.rs:4099]: libgit2 doesn't require git CLI.

```suggestion
#[cargo_test(public_network_test)]
```

File: src/cargo/sources/git/utils.rs [hunk 6]
```diff
@@ -1011,152 +1011,9 @@ pub fn fetch(
...
-                    if reinitialize(repo).is_ok() {
-                        continue;
-                    }
-                }
-
-                return Err(err.into());
-            }
-            Ok(())
-        })
+        return fetch_with_libgit2(repo, remote_url, refspecs, tags, shallow, gctx);
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs]: `return` is redundant.

File: src/cargo/sources/git/utils.rs [hunk 7]
```diff
@@ -958,6 +964,8 @@ pub fn fetch(
 
     clean_repo_temp_files(repo);
 
+    let mut revision: String = String::new();
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:967]: Probably don't need this extra variable. `oid_to_fetch` already contains the revision information we want.

File: src/cargo/sources/git/utils.rs [hunk 8]
```diff
@@ -1007,13 +1016,36 @@ pub fn fetch(
     }
 
     if let Some(true) = gctx.net_config()?.git_fetch_with_cli {
-        return fetch_with_cli(repo, remote_url, &refspecs, tags, gctx);
+        match fetch_with_cli(repo, remote_url, &refspecs, tags, gctx) {
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1019]: A better way to handling this is

```rust
let result = if let Some(true) = gctx.net_config()?.git_fetch_with_cli {
  // ..
} else if gctx.cli_unstable().gitoxide.map_or(false, |git| git.fetch) {
  // ..
} else {
  // ..
};

if fast_path_rev
  result.with_context(|| format("revision {oid_to_fetch} not found")
} else {
  result
}
```

File: src/cargo/sources/git/utils.rs [hunk 9]
```diff
@@ -1007,13 +1016,36 @@ pub fn fetch(
...
+            }
+        };
     }
 
     if gctx.cli_unstable().gitoxide.map_or(false, |git| git.fetch) {
-        return fetch_with_gitoxide(repo, remote_url, refspecs, tags, shallow, gctx);
+        match fetch_with_gitoxide(repo, remote_url, refspecs, tags, shallow, gctx) {
+            Ok(result) => return Ok(result),
+            Err(e) => {
+                if fast_path_rev == true {
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1031]: if it is a boolean, just write `if bool` directly . Don't compare `if bool == true`.

File: src/cargo/sources/git/utils.rs [hunk 10]
```diff
@@ -1007,13 +1016,36 @@ pub fn fetch(
...
+                    return Err(e).context(format!("Revision {:?} not found", revision));
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1032]: `.context` is eagerly evaluated — a string of `Revision ...` would be allocated even when it is not on the error path. use `with_context` to avoid that.

Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1032]: In Cargo by convention we use lowercase for the first character for the first sentence of a message.

File: src/cargo/sources/git/utils.rs [hunk 11]
```diff
@@ -1227,6 +1084,171 @@ fn fetch_with_cli(
...
+}
+
+fn fetch_with_libgit2(
+    repo: &mut git2::Repository,
+    remote_url: &str,
+    refspecs: Vec<String>,
+    tags: bool,
+    shallow: gix::remote::fetch::Shallow,
+    gctx: &GlobalContext,
+) -> Result<(), anyhow::Error> {
```
Comment (weihanglo) [inline src/cargo/sources/git/utils.rs:1234]: In Cargo  `CargoResult<T>` is used as a synonym of `Result<T, anyhow::Error>`. You're not wrong here just we have a subtle convention.

```suggestion
) -> CargoResult<()> {
```

File: tests/testsuite/git.rs [hunk 12]
```diff
@@ -4052,6 +4052,100 @@ src/lib.rs
...
+Caused by:
+  revision 11111b376b93484341c68fbca3ca110ae5cd2790 not found
+
+Caused by:
+  network failure seems to have happened
+  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
+  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
+
+Caused by:
+  could not read from remote repository; class=Net (12); code=Eof (-20)
```
Comment (weihanglo) [inline tests/testsuite/git.rs:4143]: It is interesting that the error message doesn't seem to be emitted on GitHub Actions machine, but only on my local machine.

If it still happens, instead of asserting the entire message, try omitting some parts we're not interested in with `...` (For the syntax, see https://doc.rust-lang.org/nightly/nightly-rustc/cargo_test_support/compare/index.html).


```suggestion
[UPDATING] git repository `https://github.com/rust-lang/bitflags.git`
...
[ERROR] failed to get `bitflags` as a dependency of package `foo v0.1.0 ([ROOT]/foo)`
Caused by:
  failed to load source for dependency `bitflags`
Caused by:
  Unable to update https://github.com/rust-lang/bitflags.git?rev=11111b376b93484341c68fbca3ca110ae5cd2790
Caused by:
  failed to clone into: [ROOT]/home/.cargo/git/db/bitflags-[HASH]
Caused by:
  revision 11111b376b93484341c68fbca3ca110ae5cd2790 not found
...
```

Comment (weihanglo) [reply tests/testsuite/git.rs:4143]: For the context:

```
---- git::git_fetch_libgit2_error_message stdout ----
running `/Users/runner/work/cargo/cargo/target/debug/cargo fetch`
thread 'git::git_fetch_libgit2_error_message' panicked at tests/testsuite/git.rs:4163:10:

---- expected: tests/testsuite/git.rs:4135:27
++++ actual:   stderr
   1    1 | [UPDATING] git repository `[https://github.com/rust-lang/bitflags.git`](https://github.com/rust-lang/bitflags.git%60)
   2      - [WARNING] spurious network error (3 tries remaining): could not read from remote repository; class=Net (12); code=Eof (-20)
   3      - [WARNING] spurious network error (2 tries remaining): could not read from remote repository; class=Net (12); code=Eof (-20)
   4      - [WARNING] spurious network error (1 tries remaining): could not read from remote repository; class=Net (12); code=Eof (-20)
   5    2 | [ERROR] failed to get `bitflags` as a dependency of package `foo v0.1.0 ([ROOT]/foo)`
   6    3 | 
   7    4 | Caused by:
   8    5 |   failed to load source for dependency `bitflags`
   9    6 | 
          ⋮
  15   12 | 
  16   13 | Caused by:
  17   14 |   revision 11111b376b93484341c68fbca3ca110ae5cd2790 not found
  18   15 | 
  19   16 | Caused by:
  20      -   network failure seems to have happened
  21      -   if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  22      -   https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
       17 +   Could not decode server reply
  23   18 | 
  24   19 | Caused by:
  25      -   could not read from remote repository; class=Net (12); code=Eof (-20)
       20 +   Unknown or unsupported header: "ERR upload-pack: not our ref 11111b376b93484341c68fbca3ca110ae5cd2790"
       ```
